sale_order_type/models/sale_order.py

```diff
@@ -53,21 +53,30 @@ def onchange_type_id(self):
 
     @api.model
     def create(self, vals):
-        if vals.get('name', '/') == '/'and vals.get('type_id'):
+        if vals.get('name', '/') == '/' and vals.get('type_id'):
             sale_type = self.env['sale.order.type'].browse(vals['type_id'])
             if sale_type.sequence_id:
                 vals['name'] = sale_type.sequence_id.next_by_id()
         return super(SaleOrder, self).create(vals)
 
     @api.multi
     def write(self, vals):
-        res = super(SaleOrder, self).write(vals)
-        if vals.get("type_id") and (self.state == "draft" or self.state == "sent"):
-            for order in self:
-                sale_type = self.env["sale.order.type"].browse(vals["type_id"])
-                if sale_type.sequence_id:
-                    order.name = sale_type.sequence_id.next_by_id()
-        return res
+        if vals.get("type_id"):
+            sale_type = self.env["sale.order.type"].browse(vals["type_id"])
+            if sale_type.sequence_id:
+                for record in self:
+                    if (
+                        record.state in {"draft", "sent"}
+                        and record.type_id != sale_type
+                    ):
+                        new_vals = vals.copy()
+                        new_vals["name"] = sale_type.sequence_id.next_by_id()
+                        super(SaleOrder, record).write(new_vals)
+                    else:
+                        super(SaleOrder, record).write(vals)
+            return True
+        else:
+            return super(SaleOrder, record).write(vals)
```

Comment:
Sorry again, but I'm detecting a problem, and it's if there's a new type, but no sequence: no write will be done. Proper code is:
```suggestion
        if vals.get("type_id"):
            sale_type = self.env["sale.order.type"].browse(vals["type_id"])
            if sale_type.sequence_id:
                for record in self:
                    if (
                        record.state in {"draft", "sent"}
                        and record.type_id != sale_type
                    ):
                        new_vals = vals.copy()
                        new_vals["name"] = sale_type.sequence_id.next_by_id()
                        super(SaleOrder, record).write(new_vals)
                    else:
                        super(SaleOrder, record).write(vals)
                return True
        return super(SaleOrder, record).write(vals)
```

Reply:
got it ..